metadata-ingestion/src/datahub/ingestion/source/dremio/dremio_source.py

```diff
@@ -0,0 +1,616 @@
+                        continue
+
+                    datahub_source_type = (
+                        self.reference_source_mapping.get_datahub_source_type(
+                            source_type
+                        )
+                    )
+
+                    if re.search(mapping.platform, datahub_source_type, re.IGNORECASE):
+                        source_platform_name = source_platform_name.lower()
+                        source_map[source_platform_name] = mapping
+                        source_map[
+                            source_platform_name
+                        ].dremio_source_type = self.reference_source_mapping.get_category(
+                            source_type
+                        )
+                        source_map[source_platform_name].root_path = root_path
+                        source_map[source_platform_name].database_name = database_name
+                        source_present = True
+                        break
+
+                if not source_present:
+                    try:
+                        dremio_source_type = self.reference_source_mapping.get_category(
+                            source_type
+                        )
+                    except Exception as exc:
+                        logger.info(
+                            f"Source {source_type} is not a standard Dremio source type. "
+                            f"Adding source_type {source_type} to mapping as database. Error: {exc}"
+                        )
+
+                        self.reference_source_mapping.add_mapping(
+                            source_type, source_name
+                        )
+                        dremio_source_type = self.reference_source_mapping.get_category(
+                            source_type
+                        )
+
+                    source_map[source_platform_name.lower()] = DremioSourceMapping(
+                        platform=source_type,
+                        platform_name=source_name,
+                        dremio_source_type=dremio_source_type,
+                    )
+
+            else:
+                logger.error(
+                    f'Source "{source.container_name}" is broken. Containers will not be created for source.'
+                )
+                logger.error(
+                    f'No new cross-platform lineage will be emitted for source "{source.container_name}".'
+                )
+                logger.error("Fix this source in Dremio to fix this issue.")
+
+        return source_map
+
+    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Generate workunits for Dremio metadata.
+        """
+
+        self.source_map = self._build_source_map()
+
+        for wu in self.get_workunits_internal():
+            self.report.report_workunit(wu)
+            yield wu
+
+        # Emit the stale entity removal workunits
+        yield from self.stale_entity_removal_handler.gen_removed_entity_workunits()

```

Comment:
when stale entity removal workunit processor is used, `get_workunits` can be removed in favour of using only `get_workunits_internal` . It should also auto report workunits. Please move build_source_map call to get_workunits_internal.